native_client/beam_search.cc

```diff
@@ -0,0 +1,546 @@
+    }
+
+    auto seq_len_t = (*seq_len)->vec<tf::int32>();
+
+    for (int b = 0; b < batch_size; ++b) {
+      if (!(seq_len_t(b) <= max_time)) {
+        return tf::errors::FailedPrecondition("sequence_length(", b, ") <= ",
+                                          max_time);
+      }
+    }
+
+    tf::Status s = ctx->allocate_output(
+        "log_probability", tf::TensorShape({batch_size, top_paths_}), log_prob);
+    if (!s.ok()) return s;
+
+    s = ctx->output_list("decoded_indices", decoded_indices);
+    if (!s.ok()) return s;
+    s = ctx->output_list("decoded_values", decoded_values);
+    if (!s.ok()) return s;
+    s = ctx->output_list("decoded_shape", decoded_shape);
+    if (!s.ok()) return s;
+
+    return tf::Status::OK();
+  }
+
+  // sequences[b][p][ix] stores decoded value "ix" of path "p" for batch "b".
+  tf::Status StoreAllDecodedSequences(
+      const std::vector<std::vector<std::vector<int>>> &sequences,
+      tf::OpOutputList *decoded_indices, tf::OpOutputList *decoded_values,
+      tf::OpOutputList *decoded_shape) const {
+    // Calculate the total number of entries for each path
+    const tf::int64 batch_size = sequences.size();
+    std::vector<tf::int64> num_entries(top_paths_, 0);
+
+    // Calculate num_entries per path
+    for (const auto &batch_s : sequences) {
+      CHECK_EQ(batch_s.size(), top_paths_);
+      for (int p = 0; p < top_paths_; ++p) {
+        num_entries[p] += batch_s[p].size();
+      }
+    }
+
+    for (int p = 0; p < top_paths_; ++p) {
+      tf::Tensor *p_indices = nullptr;
+      tf::Tensor *p_values = nullptr;
+      tf::Tensor *p_shape = nullptr;
+
+      const tf::int64 p_num = num_entries[p];
+
+      tf::Status s =
+          decoded_indices->allocate(p, tf::TensorShape({p_num, 2}), &p_indices);
+      if (!s.ok()) return s;
+      s = decoded_values->allocate(p, tf::TensorShape({p_num}), &p_values);
+      if (!s.ok()) return s;
+      s = decoded_shape->allocate(p, tf::TensorShape({2}), &p_shape);
+      if (!s.ok()) return s;
+
+      auto indices_t = p_indices->matrix<tf::int64>();
+      auto values_t = p_values->vec<tf::int64>();
+      auto shape_t = p_shape->vec<tf::int64>();
+
+      tf::int64 max_decoded = 0;
+      tf::int64 offset = 0;
+
+      for (tf::int64 b = 0; b < batch_size; ++b) {
+        auto &p_batch = sequences[b][p];
+        tf::int64 num_decoded = p_batch.size();
+        max_decoded = std::max(max_decoded, num_decoded);
+        std::copy_n(p_batch.begin(), num_decoded, &values_t(offset));
+        for (tf::int64 t = 0; t < num_decoded; ++t, ++offset) {
+          indices_t(offset, 0) = b;
+          indices_t(offset, 1) = t;
+        }
+      }
+
+      shape_t(0) = batch_size;
+      shape_t(1) = max_decoded;
+    }
+    return tf::Status::OK();
+  }
+
+ private:
+  int top_paths_;
+  TF_DISALLOW_COPY_AND_ASSIGN(CTCDecodeHelper);
+};
+
+// CTC beam search
+class CTCBeamSearchDecoderOp : public tf::OpKernel {
```

Comment:
That class name is already used by Tensorflow vanilla CTC Beam Search decoder. Can we rename to something matching the op? Like `CTCBeamSearchDecoderWithLM`.

Reply:
Technically this is `::CTCBeamSearchDecoderOp` rather than `::tensorflow::CTCBeamSearchDecoderOp`, but you're right, I'll rename it :)